File: src/translate/validation.ts
import type { OpenAIChatRequest } from "./types.js";
import type { AnthropicRequest } from "./anthropic-types.js";

export class ValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ValidationError";
  }
}

export function validateOpenAIChatRequest(body: unknown): OpenAIChatRequest {
  if (!body || typeof body !== "object") {
    throw new ValidationError("Request body must be a JSON object");
  }

  const req = body as Record<string, unknown>;

  if (req.model !== undefined && typeof req.model !== "string") {
    throw new ValidationError("Field 'model' must be a string");
  }

  if (!Array.isArray(req.messages)) {
    throw new ValidationError("Field 'messages' must be an array");
  }

  for (let i = 0; i < req.messages.length; i++) {
    const msg = req.messages[i];
    if (!msg || typeof msg !== "object") {
      throw new ValidationError(`messages[${i}] must be an object`);
    }
    const m = msg as Record<string, unknown>;
    const validRoles = ["system", "developer", "user", "assistant", "tool"];
    if (typeof m.role !== "string" || !validRoles.includes(m.role)) {
      throw new ValidationError(`messages[${i}].role must be one of: ${validRoles.join(", ")}`);
    }
    if (m.content === undefined && m.tool_calls === undefined) {
      throw new ValidationError(`messages[${i}] must contain either 'content' or 'tool_calls'`);
    }
    if (m.role === "tool" && typeof m.tool_call_id !== "string") {
      throw new ValidationError(`messages[${i}].tool_call_id must be a string when role is "tool"`);
    }
  }

  // Top-level scalar guards — fail fast with a local 400 instead of proxying
  // a confusing upstream 400.
  if (req.temperature !== undefined && typeof req.temperature !== "number") {
    throw new ValidationError("Field 'temperature' must be a number");
  }
  if (req.temperature !== undefined && (req.temperature < 0 || req.temperature > 2)) {
    throw new ValidationError("Field 'temperature' must be between 0 and 2");
  }
  if (req.top_p !== undefined && typeof req.top_p !== "number") {
    throw new ValidationError("Field 'top_p' must be a number");
  }
  if (req.top_p !== undefined && (req.top_p < 0 || req.top_p > 1)) {
    throw new ValidationError("Field 'top_p' must be between 0 and 1");
  }
  if (req.max_tokens !== undefined && typeof req.max_tokens !== "number") {
    throw new ValidationError("Field 'max_tokens' must be a number");
  }
  if (req.max_tokens !== undefined && (!Number.isFinite(req.max_tokens) || req.max_tokens <= 0)) {
    throw new ValidationError("Field 'max_tokens' must be a positive number");
  }
  if (req.tool_choice !== undefined) {
    const tc = req.tool_choice;
    const validStrings = new Set(["auto", "none", "required"]);
    if (typeof tc === "string" && !validStrings.has(tc)) {
      throw new ValidationError(`Field 'tool_choice' string must be one of: ${[...validStrings].join(", ")}`);
    }
    if (typeof tc === "object" && tc !== null) {
      const t = tc as Record<string, unknown>;
      if (t.type !== "function") {
        throw new ValidationError(`Field 'tool_choice.type' must be "function" when tool_choice is an object`);
      }
    }
  }

  return body as OpenAIChatRequest;
}

// ── Anthropic validation ──

function isRecord(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

/** Validate only fields consumed by the local estimator, not generation fields. */
export function validateCountTokensRequest(body: unknown): Record<string, unknown> {
  if (!isRecord(body)) throw new ValidationError("Request body must be a JSON object");
  if (body.system !== undefined && typeof body.system !== "string") {
    if (
      !Array.isArray(body.system) ||
      !body.system.every((b) => isRecord(b) && (b.text === undefined || typeof b.text === "string"))
    ) {
      throw new ValidationError("Field 'system' must be a string or array of text blocks");
    }
  }
  if (body.messages !== undefined) {
    if (
      !Array.isArray(body.messages) ||
      !body.messages.every(
        (m) =>
          isRecord(m) &&
          (typeof m.content === "string" ||
            (Array.isArray(m.content) && m.content.every(isRecord))),
      )
    ) {
      throw new ValidationError(
        "Field 'messages' must contain objects with string or block-array content",
      );
    }
  }
  if (body.tools !== undefined) {
    if (
      !Array.isArray(body.tools) ||
      !body.tools.every(
        (t) =>
          isRecord(t) &&
          (t.name === undefined || typeof t.name === "string") &&
          (t.description === undefined || typeof t.description === "string") &&
          (t.input_schema === undefined || isRecord(t.input_schema)),
      )
    ) {
      throw new ValidationError("Field 'tools' must be an array of tool objects");
    }
  }
  return body;
}

const UNSUPPORTED_CONTENT_TYPES = new Set([
  "document",
  "search_result",
  "web_search_tool_result",
  "web_fetch_tool_result",
  "code_execution_tool_result",
  "mcp_tool_result",
  "container_upload",
  "server_tool_use",
  "mid_conversation_system",
]);

/** Accepted `thinking.type` values (Messages API: enabled / disabled / adaptive). */
const THINKING_TYPES = new Set(["enabled", "disabled", "adaptive"]);

/** Accepted `output_config.effort` values — the discrete levels CC accepts. */
const EFFORT_LEVELS = new Set(["low", "medium", "high", "xhigh", "max"]);

const BUILT_IN_TOOL_TYPES = new Set([
  "computer_20241022",
  "bash_20241022",
  "text_editor_20241022",
  "web_search_20250305",
]);

export function validateAnthropicRequest(body: unknown): AnthropicRequest {
  if (!body || typeof body !== "object") {
    throw new ValidationError("Request body must be a JSON object");
  }

  const req = body as Record<string, unknown>;

  if (typeof req.model !== "string") {
    throw new ValidationError("Field 'model' must be a string");
  }

  if (typeof req.max_tokens !== "number") {
    throw new ValidationError("Field 'max_tokens' must be a number");
  }

  if (!Array.isArray(req.messages)) {
    throw new ValidationError("Field 'messages' must be an array");
  }

  for (let i = 0; i < req.messages.length; i++) {
    const msg = req.messages[i] as Record<string, unknown>;
    if (msg.role !== "user" && msg.role !== "assistant" && msg.role !== "system") {
      throw new ValidationError(`messages[${i}].role must be "user" or "assistant"`);
    }
    if (msg.content === undefined) {
      throw new ValidationError(`messages[${i}] must contain 'content'`);
    }
    validateContentBlocks(msg.content, i);
  }

  if (Array.isArray(req.tools)) {
    for (let i = 0; i < req.tools.length; i++) {
      const tool = req.tools[i] as Record<string, unknown>;
      const type = tool.type as string | undefined;
      if (type && BUILT_IN_TOOL_TYPES.has(type)) {
        throw new ValidationError(
          `tools[${i}]: built-in tool type "${type}" is not supported. Only custom tools ({name, description, input_schema}) are allowed.`,
        );
      }
    }
  }

  if (req.thinking && typeof req.thinking === "object") {
    const t = req.thinking as Record<string, unknown>;
    // The Messages API defines three values; a client's "thinking off" setting
    // sends "disabled" and its adaptive setting sends "adaptive". Rejecting
    // anything but "enabled" turned a normal setting into a 400 that looked
    // like a connectivity failure downstream.
    if (t.type !== undefined && !THINKING_TYPES.has(t.type as string)) {
      throw new ValidationError(
        `Field 'thinking.type' must be one of: ${[...THINKING_TYPES].join(", ")}`,
      );
    }
    // budget_tokens only exists on the "enabled" form, and is only meaningful
    // there; guard it against max_tokens so the upstream doesn't 400 later.
    if (typeof t.budget_tokens === "number" && t.budget_tokens >= (req.max_tokens as number)) {
      throw new ValidationError("thinking.budget_tokens must be less than max_tokens");
    }
  }

  if (req.output_config !== undefined) {
    const oc = req.output_config as Record<string, unknown>;
    if (oc === null || typeof oc !== "object") {
      throw new ValidationError("Field 'output_config' must be an object");
    }
    if (oc.effort !== undefined && !EFFORT_LEVELS.has(oc.effort as string)) {
      throw new ValidationError(
        `Field 'output_config.effort' must be one of: ${[...EFFORT_LEVELS].join(", ")}`,
      );
    }
  }

  // Top-level scalar guards (Anthropic)
  if (req.temperature !== undefined && typeof req.temperature !== "number") {
    throw new ValidationError("Field 'temperature' must be a number");
  }
  if (req.top_p !== undefined && typeof req.top_p !== "number") {
    throw new ValidationError("Field 'top_p' must be a number");
  }
  if (req.top_k !== undefined && typeof req.top_k !== "number") {
    throw new ValidationError("Field 'top_k' must be a number");
  }

  return body as AnthropicRequest;
}

function validateContentBlocks(content: unknown, msgIdx: number): void {
  if (typeof content === "string") return;
  if (!Array.isArray(content)) {
    throw new ValidationError(`messages[${msgIdx}].content must be a string or array of blocks`);
  }
  for (let i = 0; i < content.length; i++) {
    const block = content[i] as Record<string, unknown>;
    if (typeof block.type !== "string") continue;
    if (UNSUPPORTED_CONTENT_TYPES.has(block.type)) {
      throw new ValidationError(
        `messages[${msgIdx}].content[${i}]: block type "${block.type}" is not supported`,
      );
    }
  }
}

File: crates/ccproxy/src/validation.rs
//! Request validation, ported from src/translate/validation.ts. Error strings
//! are part of the contract (the golden pins them verbatim), including the two
//! asymmetries: a missing `model` is NOT rejected locally (it is forwarded and
//! CC's 400 is passed through), and `max_tokens` is optional for OpenAI but
//! required for Anthropic.

use serde_json::{Map, Value};

#[derive(Debug)]
pub struct ValidationError(pub String);

type Result<T> = std::result::Result<T, ValidationError>;

fn err<T>(msg: impl Into<String>) -> Result<T> {
    Err(ValidationError(msg.into()))
}

fn as_object(body: &Value) -> Option<&Map<String, Value>> {
    body.as_object()
}

/// `typeof x === "number"` in JS terms: any JSON number.
fn is_number(v: &Value) -> bool {
    v.is_number()
}

// ── OpenAI ────────────────────────────────────────────────

const OPENAI_ROLES: [&str; 5] = ["system", "developer", "user", "assistant", "tool"];

pub fn validate_openai_chat_request(body: &Value) -> Result<()> {
    let Some(req) = as_object(body) else {
        return err("Request body must be a JSON object");
    };

    if let Some(model) = req.get("model") {
        if !model.is_string() {
            return err("Field 'model' must be a string");
        }
    }

    let Some(messages) = req.get("messages").and_then(Value::as_array) else {
        return err("Field 'messages' must be an array");
    };

    for (i, msg) in messages.iter().enumerate() {
        let Some(m) = as_object(msg) else {
            return err(format!("messages[{i}] must be an object"));
        };
        match m.get("role").and_then(Value::as_str) {
            Some(role) if OPENAI_ROLES.contains(&role) => {}
            _ => {
                return err(format!(
                    "messages[{i}].role must be one of: {}",
                    OPENAI_ROLES.join(", ")
                ))
            }
        }
        if !m.contains_key("content") && !m.contains_key("tool_calls") {
            return err(format!(
                "messages[{i}] must contain either 'content' or 'tool_calls'"
            ));
        }
        if m.get("role").and_then(Value::as_str) == Some("tool")
            && !m.get("tool_call_id").is_some_and(Value::is_string)
        {
            return err(format!(
                "messages[{i}].tool_call_id must be a string when role is \"tool\""
            ));
        }
    }

    if let Some(t) = req.get("temperature") {
        if !is_number(t) {
            return err("Field 'temperature' must be a number");
        }
        let n = t.as_f64().unwrap_or(0.0);
        if !(0.0..=2.0).contains(&n) {
            return err("Field 'temperature' must be between 0 and 2");
        }
    }
    if let Some(p) = req.get("top_p") {
        if !is_number(p) {
            return err("Field 'top_p' must be a number");
        }
        let n = p.as_f64().unwrap_or(0.0);
        if !(0.0..=1.0).contains(&n) {
            return err("Field 'top_p' must be between 0 and 1");
        }
    }
    if let Some(mt) = req.get("max_tokens") {
        if !is_number(mt) {
            return err("Field 'max_tokens' must be a number");
        }
        let n = mt.as_f64().unwrap_or(0.0);
        if !n.is_finite() || n <= 0.0 {
            return err("Field 'max_tokens' must be a positive number");
        }
    }
    if let Some(tc) = req.get("tool_choice") {
        if let Some(s) = tc.as_str() {
            if !["auto", "none", "required"].contains(&s) {
                return err("Field 'tool_choice' string must be one of: auto, none, required");
            }
        } else if let Some(obj) = tc.as_object() {
            if obj.get("type").and_then(Value::as_str) != Some("function") {
                return err(
                    "Field 'tool_choice.type' must be \"function\" when tool_choice is an object",
                );
            }
        }
    }

    Ok(())
}

// ── Anthropic ─────────────────────────────────────────────

const UNSUPPORTED_CONTENT_TYPES: [&str; 9] = [
    "document",
    "search_result",
    "web_search_tool_result",
    "web_fetch_tool_result",
    "code_execution_tool_result",
    "mcp_tool_result",
    "container_upload",
    "server_tool_use",
    "mid_conversation_system",
];

const THINKING_TYPES: [&str; 3] = ["enabled", "disabled", "adaptive"];
const EFFORT_LEVELS: [&str; 5] = ["low", "medium", "high", "xhigh", "max"];
const EFFORT_OFF_VALUES: [&str; 4] = ["off", "none", "disabled", "minimal"];
const BUILT_IN_TOOL_TYPES: [&str; 4] = [
    "computer_20241022",
    "bash_20241022",
    "text_editor_20241022",
    "web_search_20250305",
];

/// Whether an effort value means "no extended thinking" rather than a level.
pub fn is_effort_off(value: &Value) -> bool {
    match value.as_str() {
        Some(s) => {
            let t = s.trim().to_ascii_lowercase();
            EFFORT_OFF_VALUES.contains(&t.as_str())
        }
        None => false,
    }
}

pub fn validate_anthropic_request(body: &Value) -> Result<()> {
    let Some(req) = as_object(body) else {
        return err("Request body must be a JSON object");
    };

    if !req.get("model").is_some_and(Value::is_string) {
        return err("Field 'model' must be a string");
    }
    if !req.get("max_tokens").is_some_and(is_number) {
        return err("Field 'max_tokens' must be a number");
    }

    let Some(messages) = req.get("messages").and_then(Value::as_array) else {
        return err("Field 'messages' must be an array");
    };

    for (i, msg) in messages.iter().enumerate() {
        let Some(m) = as_object(msg) else {
            return err(format!(
                "messages[{i}].role must be \"user\" or \"assistant\""
            ));
        };
        match m.get("role").and_then(Value::as_str) {
            Some("user") | Some("assistant") | Some("system") => {}
            _ => {
                return err(format!(
                    "messages[{i}].role must be \"user\" or \"assistant\""
                ))
            }
        }
        let Some(content) = m.get("content") else {
            return err(format!("messages[{i}] must contain 'content'"));
        };
        validate_content_blocks(content, i)?;
    }

    if let Some(tools) = req.get("tools").and_then(Value::as_array) {
        for (i, tool) in tools.iter().enumerate() {
            if let Some(t) = tool.get("type").and_then(Value::as_str) {
                if BUILT_IN_TOOL_TYPES.contains(&t) {
                    return err(format!(
                        "tools[{i}]: built-in tool type \"{t}\" is not supported. Only custom tools ({{name, description, input_schema}}) are allowed."
                    ));
                }
            }
        }
    }

    if let Some(thinking) = req.get("thinking").and_then(Value::as_object) {
        if let Some(t) = thinking.get("type") {
            match t.as_str() {
                Some(s) if THINKING_TYPES.contains(&s) => {}
                _ => {
                    return err(format!(
                        "Field 'thinking.type' must be one of: {}",
                        THINKING_TYPES.join(", ")
                    ))
                }
            }
        }
        if let Some(budget) = thinking.get("budget_tokens").and_then(Value::as_f64) {
            let max_tokens = req.get("max_tokens").and_then(Value::as_f64).unwrap_or(0.0);
            if budget >= max_tokens {
                return err("thinking.budget_tokens must be less than max_tokens");
            }
        }
    }

    if let Some(oc) = req.get("output_config") {
        let Some(oc) = oc.as_object() else {
            return err("Field 'output_config' must be an object");
        };
        if let Some(effort) = oc.get("effort") {
            let lower = effort
                .as_str()
                .map(str::to_ascii_lowercase)
                .unwrap_or_default();
            if !EFFORT_LEVELS.contains(&lower.as_str()) && !is_effort_off(effort) {
                return err(format!(
                    "Field 'output_config.effort' must be one of: {}",
                    EFFORT_LEVELS.join(", ")
                ));
            }
        }
    }

    if req.get("temperature").is_some_and(|v| !is_number(v)) {
        return err("Field 'temperature' must be a number");
    }
    if req.get("top_p").is_some_and(|v| !is_number(v)) {
        return err("Field 'top_p' must be a number");
    }
    if req.get("top_k").is_some_and(|v| !is_number(v)) {
        return err("Field 'top_k' must be a number");
    }

    Ok(())
}

fn validate_content_blocks(content: &Value, msg_idx: usize) -> Result<()> {
    if content.is_string() {
        return Ok(());
    }
    let Some(blocks) = content.as_array() else {
        return err(format!(
            "messages[{msg_idx}].content must be a string or array of blocks"
        ));
    };
    for (i, block) in blocks.iter().enumerate() {
        let Some(t) = block.get("type").and_then(Value::as_str) else {
            continue;
        };
        if UNSUPPORTED_CONTENT_TYPES.contains(&t) {
            return err(format!(
                "messages[{msg_idx}].content[{i}]: block type \"{t}\" is not supported"
            ));
        }
    }
    Ok(())
}

/// Validate only the fields consumed by the local estimator.
pub fn validate_count_tokens_request(body: &Value) -> Result<()> {
    let Some(req) = body.as_object() else {
        return err("Request body must be a JSON object");
    };
    if let Some(system) = req.get("system") {
        if !system.is_string() {
            let ok = system.as_array().is_some_and(|arr| {
                arr.iter().all(|b| {
                    b.as_object()
                        .is_some_and(|o| o.get("text").is_none_or(Value::is_string))
                })
            });
            if !ok {
                return err("Field 'system' must be a string or array of text blocks");
            }
        }
    }
    if let Some(messages) = req.get("messages") {
        let ok = messages.as_array().is_some_and(|arr| {
            arr.iter().all(|m| {
                m.as_object().is_some_and(|o| match o.get("content") {
                    Some(c) if c.is_string() => true,
                    Some(c) => c.as_array().is_some_and(|a| a.iter().all(Value::is_object)),
                    None => false,
                })
            })
        });
        if !ok {
            return err("Field 'messages' must contain objects with string or block-array content");
        }
    }
    if let Some(tools) = req.get("tools") {
        let ok = tools.as_array().is_some_and(|arr| {
            arr.iter().all(|t| {
                t.as_object().is_some_and(|o| {
                    o.get("name").is_none_or(Value::is_string)
                        && o.get("description").is_none_or(Value::is_string)
                        && o.get("input_schema").is_none_or(Value::is_object)
                })
            })
        });
        if !ok {
            return err("Field 'tools' must be an array of tool objects");
        }
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn openai_missing_model_is_not_rejected_locally() {
        // Contract: this is forwarded upstream and CC's 400 is passed through.
        let body = json!({"messages": [{"role": "user", "content": "x"}]});
        assert!(validate_openai_chat_request(&body).is_ok());
    }

    #[test]
    fn openai_messages_must_be_an_array() {
        let body = json!({"model": "m", "max_tokens": 8});
        let e = validate_openai_chat_request(&body).unwrap_err();
        assert_eq!(e.0, "Field 'messages' must be an array");
    }

    #[test]
    fn openai_role_message_matches_golden() {
        let body = json!({"model":"m","max_tokens":8,"messages":[{"role":"wizard","content":"x"}]});
        let e = validate_openai_chat_request(&body).unwrap_err();
        assert_eq!(
            e.0,
            "messages[0].role must be one of: system, developer, user, assistant, tool"
        );
    }

    #[test]
    fn openai_temperature_and_tool_choice_messages() {
        let mut base = json!({"model":"m","messages":[{"role":"user","content":"x"}]});
        base["temperature"] = json!("hot");
        assert_eq!(
            validate_openai_chat_request(&base).unwrap_err().0,
            "Field 'temperature' must be a number"
        );

        let mut base = json!({"model":"m","messages":[{"role":"user","content":"x"}]});
        base["temperature"] = json!(9);
        assert_eq!(
            validate_openai_chat_request(&base).unwrap_err().0,
            "Field 'temperature' must be between 0 and 2"
        );

        let mut base = json!({"model":"m","messages":[{"role":"user","content":"x"}]});
        base["tool_choice"] = json!("whatever");
        assert_eq!(
            validate_openai_chat_request(&base).unwrap_err().0,
            "Field 'tool_choice' string must be one of: auto, none, required"
        );
    }

    #[test]
    fn anthropic_requires_model_and_max_tokens() {
        let body = json!({"messages": []});
        assert_eq!(
            validate_anthropic_request(&body).unwrap_err().0,
            "Field 'model' must be a string"
        );
        let body = json!({"model": "m", "messages": []});
        assert_eq!(
            validate_anthropic_request(&body).unwrap_err().0,
            "Field 'max_tokens' must be a number"
        );
    }

    #[test]
    fn anthropic_budget_must_be_below_max_tokens() {
        let body = json!({
            "model": "m", "max_tokens": 100, "messages": [],
            "thinking": {"type": "enabled", "budget_tokens": 100}
        });
        assert_eq!(
            validate_anthropic_request(&body).unwrap_err().0,
            "thinking.budget_tokens must be less than max_tokens"
        );
    }

    #[test]
    fn effort_off_markers_are_accepted() {
        for v in ["off", "OFF", "none", "disabled", "minimal"] {
            assert!(is_effort_off(&json!(v)), "{v} should be an off marker");
        }
        assert!(!is_effort_off(&json!("high")));
    }

    #[test]
    fn unsupported_content_block_is_rejected() {
        let body = json!({
            "model": "m", "max_tokens": 8,
            "messages": [{"role": "user", "content": [{"type": "document"}]}]
        });
        assert_eq!(
            validate_anthropic_request(&body).unwrap_err().0,
            "messages[0].content[0]: block type \"document\" is not supported"
        );
    }
}
